File: go/0144_binary_tree_preorder_traversal.go
func preorderTraversal(root *TreeNode) []int {
	res := []int{}
	if root == nil {
		return res
	}

	cur := root
	for cur != nil {
		if cur.Left == nil {
			res = append(res, cur.Val)
			cur = cur.Right
			continue
		}

		next := cur.Left
		visited := false
		for next.Right != nil {
			if next.Right == cur {
				next.Right = nil
				cur = cur.Right
				visited = true
			}
			next = next.Right
		}

		if !visited {
			next.Right = cur
			res = append(res, cur.Val)
			cur = cur.Left
		}
	}
	return res
}

func preorderTraversal(root *TreeNode) []int {
	res := []int{}
	if root == nil {
		return res
	}

	//visited := make(map[*TreeNode]bool)
	stack := []*TreeNode{}
	cur := root
	for {
		res = append(res, cur.Val)

		if cur.Left != nil {
			if cur.Right != nil {
				stack = append(stack, cur.Right)
			}
			cur = cur.Left
			continue
		}

		if cur.Right != nil {
			cur = cur.Right
			continue
		}

		if len(stack) == 0 {
			break
		}
		cur = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
	}

	return res
}

File: python/0144_binary_tree_preorder_traversal.py
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []

        cur = root
        s = []
        res = []
        while True:
            res.append(cur.val)

            if cur.left:
                if cur.right:
                    s.append(cur.right)
                cur = cur.left
            elif cur.right:
                cur = cur.right
            elif s:
                cur = s.pop()
            else:
                break
        return res
